src/conmon.c

```diff
@@ -811,24 +824,37 @@ static void container_exit_cb(G_GNUC_UNUSED GPid pid, int status, G_GNUC_UNUSED
 	}
 	container_status = status;
 	container_pid = -1;
+	/* In the case of a quickly exiting exec command, the container exit callback
+	   sometimes gets called earlier than the pid exit callback. If we quit the loop at that point
+	   we risk falsely telling the caller of conmon the runtime call failed (because runtime status
+	   wouldn't be set). Instead, don't quit the loop until runtime exit is also called, which should
+	   shortly after. */
+	if (opt_updated && create_pid > 0 && opt_exec && opt_terminal) {
+		ndebugf("container pid return handled before runtime pid return. Not quitting yet.");
+		return;
+	}
+
 	g_main_loop_quit(main_loop);
 }
 
 static void write_sync_fd(int sync_pipe_fd, int res, const char *message)
 {
 	_cleanup_free_ char *escaped_message = NULL;
 	_cleanup_free_ char *json = NULL;
+
 	const char *res_key;
+	if (opt_updated)
+		res_key = "data";
+	else if (opt_exec)
+		res_key = "exit_code";
+	else
+		res_key = "pid";
+
 	ssize_t len;
 
 	if (sync_pipe_fd == -1)
 		return;
 
-	if (opt_exec)
-		res_key = "exit_code";
-	else
-		res_key = "pid";
-
 	if (message) {
 		escaped_message = escape_json_string(message);
 		json = g_strdup_printf("{\"%s\": %d, \"message\": \"%s\"}\n", res_key, res, escaped_message);
```

Comment:
another way to solve the versioning could be to specify both `data,exit_code` or `data,pid` in the outputted JSON and have:
```
json = g_strdup_printf("{\"data\": \"%s\", \"%s\": %d, \"message\": \"%s\"}\n", res, res_key, res, escaped_message);
```

Reply:
I think this solution is much cleaner, but I am going to stick with using --api-version instead. There are a couple of other cases that --api-version covers (for instance, making sure we only send on the sync pipe once in the legacy api). --attach could be used for this, but I found it hacky and difficult to explain (which means confusing to read).

I don't think the cost of sending and formatting the longer JSON is worth it if we can't get rid of the --api-version flag.

WDYT

Reply:
yes sure, if we cannot take rid of it completely then there is no advantage in changing the JSON format